File: backend/src/main/java/org/example/coretrack/service/ProductInventoryService.java
package org.example.coretrack.service;

import java.math.BigDecimal;
import java.util.List;

import org.example.coretrack.dto.product.inventory.AddProductInventoryRequest;
import org.example.coretrack.dto.product.inventory.AddProductInventoryResponse;
import org.example.coretrack.dto.product.inventory.AllSearchInventoryResponse;
import org.example.coretrack.dto.product.inventory.InventoryTransactionResponse;
import org.example.coretrack.dto.product.inventory.ProductInventoryDetailResponse;
import org.example.coretrack.dto.product.inventory.SearchInventoryResponse;
import org.example.coretrack.dto.product.inventory.StockModifyRequest;
import org.example.coretrack.dto.product.inventory.StockSetRequest;
import org.example.coretrack.dto.product.inventory.SetMinMaxRequest;
import org.example.coretrack.dto.product.inventory.SetMinMaxResponse;
import org.example.coretrack.dto.product.inventory.BulkStockModifyRequest;
import org.example.coretrack.dto.product.inventory.BulkStockSetRequest;
import org.example.coretrack.dto.product.inventory.BulkInventoryTransactionResponse;
import org.example.coretrack.dto.product.inventory.BulkInitInventoryRequest;
import org.example.coretrack.dto.product.inventory.BulkInitInventoryResponse;
import org.example.coretrack.dto.product.inventory.InventoryEnumsResponse;
import org.example.coretrack.dto.product.inventory.TransactionEnumsResponse;
import org.example.coretrack.model.auth.User;
import org.example.coretrack.model.product.inventory.ProductInventory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ProductInventoryService {
    AddProductInventoryResponse createProductInventory(AddProductInventoryRequest request, User user);
    
    InventoryTransactionResponse setStock (Long variantId, StockSetRequest request, User user);

    InventoryTransactionResponse addStock (Long variantId, StockModifyRequest request, User user);

    InventoryTransactionResponse subtractStock (Long variantId, StockModifyRequest request, User user);
    

    Page<SearchInventoryResponse> findProduct(
            String search,
            List<String> groupProducts,
            List<String> inventoryStatus,
            Pageable pageable,
            User user);

    List<AllSearchInventoryResponse> getAllForAutocomplete (String search, User user);

    ProductInventoryDetailResponse getProductInventoryById (Long variantId, User user);

    // Bulk operations
    BulkInventoryTransactionResponse bulkSetStock(BulkStockSetRequest request, User user);
    
    BulkInventoryTransactionResponse bulkAddStock(BulkStockModifyRequest request, User user);
    
    BulkInventoryTransactionResponse bulkSubtractStock(BulkStockModifyRequest request, User user);

    // Bulk init inventory
    BulkInitInventoryResponse bulkInitInventory(BulkInitInventoryRequest request, User user);

    // Get inventory enums
    InventoryEnumsResponse getInventoryEnums();
    
    // Get inventory statuses only (for FilterBar)
    List<InventoryEnumsResponse.EnumValue> getInventoryStatuses();
    
    // Get transaction enums only (for stock operations)
    TransactionEnumsResponse getTransactionEnums();

    // Production ticket inventory integration methods
    ProductInventory addToFutureStock(Long variantId, BigDecimal quantity, User user, Long ticketId);
    ProductInventory moveFromFutureToCurrentStock(Long variantId, BigDecimal quantity, User user, Long ticketId);
    ProductInventory removeFromFutureStock(Long variantId, BigDecimal quantity, User user, Long ticketId);
    ProductInventory removeFromCurrentStock(Long variantId, BigDecimal quantity, User user, Long ticketId);
    ProductInventory removeFromAllocatedAndCurrentStock(Long variantId, BigDecimal quantity, User user, Long ticketId);
    ProductInventory addToAllocatedStock(Long variantId, BigDecimal quantity, User user, Long ticketId);
    ProductInventory removeFromAllocatedStock(Long variantId, BigDecimal quantity, User user, Long ticketId);
    ProductInventory addToCurrentStock(Long variantId, BigDecimal quantity, User user, Long ticketId);
    ProductInventory getByProductVariantId(Long variantId, User user);
    Page<SearchInventoryResponse> getAlarmProduct(String search,
            List<String> groupProducts,
            List<String> status,
            boolean sortByOldest,
            Pageable pageable,
            User user);

    // Set minimum and maximum stock levels
    SetMinMaxResponse setMinimumAlertStock(Long variantId, SetMinMaxRequest request, User user);
    
    SetMinMaxResponse setMaximumStockLevel(Long variantId, SetMinMaxRequest request, User user);
}

File: frontend/src/services/productInventoryService.ts
import apiClient from '../lib/axios';
import { toBigDecimalString } from '../lib/utils';
import {
  AddProductInventoryRequest,
  AddProductInventoryResponse,
  InventoryTransactionResponse,
  BulkStockModifyRequest,
  BulkStockSetRequest,
  BulkTransactionError,
  BulkInventoryTransactionResponse,
  BulkInitInventoryRequest,
  BulkInitError,
  BulkInitInventoryResponse,
  EnumValue,
  InventoryEnumsResponse,
  TransactionEnumsResponse,
  AllSearchInventoryResponse
} from '../types/productInventory';

export const createProductInventory = async (
  request: AddProductInventoryRequest
): Promise<AddProductInventoryResponse> => {
  const response = await apiClient.post('/product-inventory/add-inventory', request);
  return response.data;
};

// Helper function to convert number to BigDecimal string for backend
export const createInventoryRequest = (
  productVariantSku: string,
  currentStock: number,
  minAlertStock?: number,
  maxStockLevel?: number
): AddProductInventoryRequest => {
  return {
    productVariantSku,
    currentStock: toBigDecimalString(currentStock),
    minAlertStock: minAlertStock ? toBigDecimalString(minAlertStock) : undefined,
    maxStockLevel: maxStockLevel ? toBigDecimalString(maxStockLevel) : undefined,
  };
};

// Add stock to a specific variant
export const addStock = async (
  variantId: number,
  quantity: number,
  note?: string,
  referenceDocumentType?: string,
  referenceDocumentId?: number,
  transactionSource?: string
): Promise<InventoryTransactionResponse> => {
  const request = {
    newQuantity: toBigDecimalString(quantity),
    note,
    referenceDocumentType,
    referenceDocumentId,
    transactionSource,
  };
  const response = await apiClient.put(`/product-inventory/${variantId}/add`, request);
  return response.data;
};

// Subtract stock from a specific variant
export const subtractStock = async (
  variantId: number,
  quantity: number,
  note?: string,
  referenceDocumentType?: string,
  referenceDocumentId?: number,
  transactionSource?: string
): Promise<InventoryTransactionResponse> => {
  const request = {
    newQuantity: toBigDecimalString(quantity),
    note,
    referenceDocumentType,
    referenceDocumentId,
    transactionSource,
  };
  const response = await apiClient.put(`/product-inventory/${variantId}/subtract`, request);
  return response.data;
};

// Bulk operations functions
export const bulkSetStock = async (request: BulkStockSetRequest): Promise<BulkInventoryTransactionResponse> => {
  const response = await apiClient.put('/product-inventory/bulk/set', request);
  return response.data;
};

export const bulkAddStock = async (request: BulkStockModifyRequest): Promise<BulkInventoryTransactionResponse> => {
  const response = await apiClient.put('/product-inventory/bulk/add', request);
  return response.data;
};

export const bulkSubtractStock = async (request: BulkStockModifyRequest): Promise<BulkInventoryTransactionResponse> => {
  const response = await apiClient.put('/product-inventory/bulk/subtract', request);
  return response.data;
};

export const bulkInitInventory = async (request: BulkInitInventoryRequest): Promise<BulkInitInventoryResponse> => {
  const response = await apiClient.post('/product-inventory/bulk/init', request);
  return response.data;
};

export const getInventoryEnums = async (): Promise<InventoryEnumsResponse> => {
  const response = await apiClient.get('/product-inventory/enums');
  return response.data;
};

export const getInventoryStatuses = async (): Promise<EnumValue[]> => {
  const response = await apiClient.get('/product-inventory/enums/inventory-statuses');
  return response.data;
};

export const getTransactionEnums = async (): Promise<TransactionEnumsResponse> => {
  const response = await apiClient.get('/product-inventory/enums/transaction-enums');
  return response.data;
};

// Helper functions for bulk operations
export const createBulkStockSetRequest = (
  variantIds: number[],
  quantity: number,
  note?: string,
  referenceDocumentType?: string,
  referenceDocumentId?: number
): BulkStockSetRequest => {
  return {
    variantIds,
    quantity: toBigDecimalString(quantity),
    note,
    referenceDocumentType,
    referenceDocumentId,
  };
};

export const createBulkStockModifyRequest = (
  variantIds: number[],
  quantity: number,
  note?: string,
  referenceDocumentType?: string,
  referenceDocumentId?: number,
  transactionSource?: string
): BulkStockModifyRequest => {
  return {
    variantIds,
    quantity: toBigDecimalString(quantity),
    note,
    referenceDocumentType,
    referenceDocumentId,
    transactionSource,
  };
};

export const createBulkInitInventoryRequest = (
  productVariantSkus: string[],
  currentStock: number,
  minAlertStock?: number,
  maxStockLevel?: number
): BulkInitInventoryRequest => {
  return {
    productVariantSkus,
    currentStock: toBigDecimalString(currentStock),
    minAlertStock: minAlertStock ? toBigDecimalString(minAlertStock) : undefined,
    maxStockLevel: maxStockLevel ? toBigDecimalString(maxStockLevel) : undefined,
  };
};

// Product Inventory Search and Filter
export const getProductInventoryFilter = async (params: any): Promise<any> => {
  const queryParams = new URLSearchParams();
  
  if (params.search) queryParams.append('search', params.search);
  if (params.groupProducts) {
    params.groupProducts.forEach((group: string) => {
      queryParams.append('groupProducts', group);
    });
  }
  if (params.inventoryStatus) {
    params.inventoryStatus.forEach((status: string) => {
      queryParams.append('inventoryStatus', status);
    });
  }
  if (params.page !== undefined) queryParams.append('page', params.page.toString());
  if (params.size !== undefined) queryParams.append('size', params.size.toString());
  if (params.sort) queryParams.append('sort', params.sort);

  const response = await apiClient.get(`/product-inventory/filter?${queryParams.toString()}`);
  return response.data;
};

// Product Inventory Alarm Filter
export const getProductAlarmFilter = async (params: any): Promise<any> => {
  const queryParams = new URLSearchParams();
  
  if (params.search) queryParams.append('search', params.search);
  if (params.groupProducts) {
    params.groupProducts.forEach((group: string) => {
      queryParams.append('groupProducts', group);
    });
  }
  if (params.status) {
    params.status.forEach((status: string) => {
      queryParams.append('status', status);
    });
  }
  if (params.sortByOldest !== undefined) queryParams.append('sortByOldest', params.sortByOldest.toString());
  if (params.page !== undefined) queryParams.append('page', params.page.toString());
  if (params.size !== undefined) queryParams.append('size', params.size.toString());

  const response = await apiClient.get(`/product-inventory/alarm/filter?${queryParams.toString()}`);
  return response.data;
};

export const getAllProductInventoryForAutocomplete = async (search?: string): Promise<AllSearchInventoryResponse[]> => {
  const params = search ? `?search=${encodeURIComponent(search)}` : '';
  const response = await apiClient.get(`/product-inventory/autocomplete${params}`);
  return response.data;
};

export const getProductInventoryById = async (variantId: number): Promise<any> => {
  const response = await apiClient.get(`/product-inventory/${variantId}`);
  return response.data;
};

 